compiler/cpp/src/thrift/generate/t_py_generator.cc

```diff
@@ -888,12 +888,33 @@ void t_py_generator::generate_py_struct_definition(ostream& out,
 
   if (is_immutable(tstruct)) {
     out << endl;
-    out << indent() << "def __setattr__(self, *args):" << endl
-        << indent() << indent_str() << "raise TypeError(\"can't modify immutable instance\")" << endl
-        << endl;
-    out << indent() << "def __delattr__(self, *args):" << endl
-        << indent() << indent_str() << "raise TypeError(\"can't modify immutable instance\")" << endl
-        << endl;
+    out << indent() << "def __setattr__(self, attr, value):" << endl;
+    indent_up();
+    if (gen_slots_) {
+      out << indent() << "if attr not in self.__slots__:" << endl;
+    } else {
+      out << indent() << "if not self.__dict__.has_key(attr):" << endl;
```

Comment:
this might not work as expected:
```
$ python3
Python 3.11.4 (main, Jun  7 2023, 10:13:09) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo:
...   _bar = None
... 
>>> foo = Foo()
>>> foo.__dict__
{}
>>> foo._bar = 1
>>> foo.__dict__
{'_bar': 1}
```
it seems that the `__dict__` would only have the key if it's used at least once?

Reply:
Currently, it looks like [each use defined field](https://github.com/apache/thrift/pull/2816/files#diff-ad110290fb2055999f700940287b9f8db58007832084247e6ffb0bd2b14b897dR850-R887) is explicitly set in the `__init__`. I believe this would avoid this issue. Is that not correct?

Alternately I could keep track of user defined fields similar to how we do for `__slots__`, thoughts?

Reply:
my main concern here is that after contextmanager changes the `__traceback__` it will start to appear in `__dict__` and then it can no longer be changed again

Reply:
Long story short, `__traceback__` does not appear in `__dict__` from what I could track down it's because the implementation of `Exception` is in C and `__traceback__` is not technically an official property.

Fishy and I discussed off PR and decided that it was acceptable to only support context managers (in Python 3.11) with when using slots. This means any user that runs into this issue should use slots to get around their issue.